worker/uniter/relation/resolver.go

```diff
@@ -72,42 +74,80 @@ func (r *relationsResolver) NextOp(localState resolver.LocalState, remoteState r
 		return nil, errors.Trace(err)
 	}
 
-	// Check whether we need to fire a hook for any of the relations
-	for relationId, relationSnapshot := range remoteState.Relations {
-		if !r.stateTracker.IsKnown(relationId) {
-			r.logger.Tracef("unknown relation %d resolving next op", relationId)
-			continue
-		} else if isImplicit, _ := r.stateTracker.IsImplicit(relationId); isImplicit {
+	// Collect peer relations and defer their processing until after other
+	// relations. This is simpler than implementing a sort based on the type.
+	// Processing them last ensures that upon application removal, the hooks
+	// for other relations can rely on the peer relations still being present.
+	var peerRelations []int
+
+	// Check whether we need to fire a hook for any of the non-peer relations.
+	for relationID, relationSnapshot := range remoteState.Relations {
+		if isPeer, _ := r.stateTracker.IsPeerRelation(relationID); isPeer {
+			peerRelations = append(peerRelations, relationID)
 			continue
 		}
 
-		// If either the unit or the relation are Dying, or the
-		// relation becomes suspended, then the relation should be
-		// broken.
-		var remoteBroken bool
-		if remoteState.Life == life.Dying || relationSnapshot.Life == life.Dying || relationSnapshot.Suspended {
-			relationSnapshot = remotestate.RelationSnapshot{}
-			remoteBroken = true
-			// TODO(axw) if relation is implicit, leave scope & remove.
+		op, err := r.processRelationSnapshot(relationID, relationSnapshot, remoteState, opFactory)
+		if err != nil {
+			if errors.Is(err, resolver.ErrNoOperation) {
+				continue
+			}
+			return nil, errors.Trace(err)
 		}
+		return op, nil
+	}
 
-		// Examine local/remote states and figure out if a hook needs
-		// to be fired for this relation.
-		relState, err := r.stateTracker.State(relationId)
+	// Process the deferred peer relations.
+	for _, relationID := range peerRelations {
+		relationSnapshot := remoteState.Relations[relationID]
+
+		op, err := r.processRelationSnapshot(relationID, relationSnapshot, remoteState, opFactory)
 		if err != nil {
-			//
-			relState = NewState(relationId)
-		}
-		hInfo, err := r.nextHookForRelation(relState, relationSnapshot, remoteBroken)
-		if err == resolver.ErrNoOperation {
-			continue
+			if errors.Is(err, resolver.ErrNoOperation) {
+				continue
+			}
+			return nil, errors.Trace(err)
 		}
-		return opFactory.NewRunHook(hInfo)
+		return op, nil
 	}
 
 	return nil, resolver.ErrNoOperation
 }
 
+func (r *relationsResolver) processRelationSnapshot(
```

Comment:
Not essential, but could be good to have a comment on the function/rename it. The name `processRelationSnapshot` tells me nothing. Would rather not have to read the comments in the body to work out whats going on